tap_kit/streams.py

```
@@ -148,3 +152,21 @@ def update_start_date_bookmark(self):
     def update_and_return_bookmark(self):
         self.update_start_date_bookmark()
         return self.get_bookmark()
+
+def validate_ingestible_data(record):
+    from pycore.text import to_unicode
+    import re
+    
+    for key, value in record.items():
+        if isinstance(value, dict):
+            validate_ingestible_data(value)
+        else:
+            try:
+                record[key] = to_unicode(value)
+            except UnicodeDecodeError as e:
+                record[key] = ''
+                pass
+            if len(str(value)) != len(str(value).encode()):
+                record[key] = value.encode("ascii", "ignore").replace(b'\x00', b'').decode()
```

if we're just trying to remove non utf-8 chars can't we encode to utf-8, ignore errors, then decode?  then we can skip the replace part.

---

When testing just encode then decode, the NUL byte persists. That's why I added the replace piece so when I check for '\u0000' in the value, it is no longer there

---

NUL is a valid uf-8 value.  Is this causing an issue downstream?

---

Yes - based on the ticket @hsessa said: "failing ungracefully at the load-reference-tables-contact_facts level with _csv.Error: line contains NUL"

---

ah ya I don't think that would be caused by null bytes.  that's probably due to foreign characters or something else weird.  I think if we just force utf-8 via `.decode('utf-8','ignore').encode("utf-8")` that would probably cover it.

---

Joel and I discussed and he found CSVs just do not support nul bytes so cleaned up the code to only strip out nulls